<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>CompGenR</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="compgenR_files/libs/clipboard/clipboard.min.js"></script>
<script src="compgenR_files/libs/quarto-html/quarto.js"></script>
<script src="compgenR_files/libs/quarto-html/popper.min.js"></script>
<script src="compgenR_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="compgenR_files/libs/quarto-html/anchor.min.js"></script>
<link href="compgenR_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="compgenR_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="compgenR_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="compgenR_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="compgenR_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">CompGenR</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="quarto" class="level2">
<h2 class="anchored" data-anchor-id="quarto">Quarto</h2>
<p>Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto see <a href="https://quarto.org" class="uri">https://quarto.org</a>.</p>
</section>
<section id="running-code" class="level2">
<h2 class="anchored" data-anchor-id="running-code">Running Code</h2>
<p>When you click the <strong>Render</strong> button a document will be generated that includes both content and the output of embedded code. You can embed code like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">+</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>You can add options to executable code like this</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>The <code>echo: false</code> option disables the printing of code (only output is displayed).</p>
</section>
<section id="computational-genomics-with-r" class="level1">
<h1>Computational Genomics with R</h1>
<p>This is my walkthrough of the book computational genomics with R by Altuna Alkalin, which can be found <a href="https://compgenomr.github.io/book/">here</a>.</p>
<section id="compute-resources" class="level3">
<h3 class="anchored" data-anchor-id="compute-resources">Compute resources</h3>
<p>I’m starting out with the initial allocation from Posit.</p>
<p>1GB Ram, 1 CPU</p>
<p>We’ll see if I need to / can bump that up if needed.</p>
</section>
<section id="installations" class="level3">
<h3 class="anchored" data-anchor-id="installations">Installations</h3>
<p>First install all of the packages. Code copied from the preface of the book. Let’s hope it works. That’s a lot of packages.</p>
<p>We also need the package that comes with the book. This contains data files needed for the analyses in the book. Files in the package are accessed using the system.file() function.</p>
<p>Now load the devtools package</p>
<p>Now try installing the compgenomRpackage</p>
<p>Seems to have worked so far</p>
<p>Let’s try reading in some files from the package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>enhancerFilePath<span class="ot">=</span><span class="fu">system.file</span>(<span class="st">"extdata"</span>,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                <span class="st">"subset.enhancers.hg18.bed"</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">package=</span><span class="st">"compGenomRData"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>cpgiFilePath<span class="ot">=</span><span class="fu">system.file</span>(<span class="st">"extdata"</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"subset.cpgi.hg18.bed"</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">package=</span><span class="st">"compGenomRData"</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># read enhancer marker BED file</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>enh.df <span class="ot">&lt;-</span> <span class="fu">read.table</span>(enhancerFilePath, <span class="at">header =</span> <span class="cn">FALSE</span>) </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co"># read CpG island BED file</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>cpgi.df <span class="ot">&lt;-</span> <span class="fu">read.table</span>(cpgiFilePath, <span class="at">header =</span> <span class="cn">FALSE</span>) </span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="co"># check first lines to see how the data looks like</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(enh.df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     V1     V2     V3 V4   V5 V6    V7      V8 V9
1 chr20 266275 267925  . 1000  .  9.11 13.1693 -1
2 chr20 287400 294500  . 1000  . 10.53 13.0231 -1
3 chr20 300500 302500  . 1000  .  9.10 13.3935 -1
4 chr20 330400 331800  . 1000  .  6.39 13.5105 -1
5 chr20 341425 343400  . 1000  .  6.20 12.9852 -1
6 chr20 437975 439900  . 1000  .  6.31 13.5184 -1</code></pre>
</div>
</div>
<p>Nice!</p>
<p>Other ways to read in files is shown in the code below. The code is commented out so that is doesn’t run.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#library(data.table)</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">#df.f=d(enhancerFilePath, header = FALSE,data.table=FALSE)</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#library(readr)</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">#df.f2=read_table(enhancerFilePath, col_names = FALSE)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Chapters 2 and 3 go over some basic R functionality. I’m skipping that here and jumping to chapter 4 because that starts analyzing the genomic data. I will return to chs 2 and 3 in a separate document.</p>
</section>
<section id="chapter-4-exploratory-data-analysis-with-unsupervised-machine-learning" class="level2">
<h2 class="anchored" data-anchor-id="chapter-4-exploratory-data-analysis-with-unsupervised-machine-learning"><strong>Chapter 4</strong> Exploratory Data Analysis with Unsupervised Machine Learning</h2>
<section id="clustering" class="level3">
<h3 class="anchored" data-anchor-id="clustering">4.1 Clustering</h3>
<p>They use a simple table for the following illustration. I reproduced that as simple_patient_data and uploaded that file. Now let’s read that into R as a dataframe called pd for patient data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Installing</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"readr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into '/cloud/lib/x86_64-pc-linux-gnu-library/4.3'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loading</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"readr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load the readxl library and import the dtaset</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>pd <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"simple_patient_data.csv"</span>, <span class="at">header =</span><span class="cn">TRUE</span>, <span class="at">row.names   =</span> <span class="dv">1</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co">#View(pd)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s see how we can calculate distances between the patients in R. We will use three distance methods.</p>
<ul>
<li><p>the&nbsp;<strong>“Manhattan distance”</strong>&nbsp;or&nbsp;<strong>“L1 norm”</strong>.</p></li>
<li><p><strong>“Euclidean Distance”</strong>&nbsp;or&nbsp;<strong>“L2 norm”</strong></p></li>
<li><p>the&nbsp;<strong>“correlation distance”</strong></p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#first the manhattan method of the dist function.</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dist</span>(pd,<span class="at">method=</span><span class="st">"manhattan"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Patient 1 Patient 2 Patient 3
Patient 2         7                    
Patient 3        24        27          
Patient 4        25        28         3</code></pre>
</div>
</div>
<p>Now try the Euclidean method</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dist</span>(pd,<span class="at">method=</span><span class="st">"euclidean"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Patient 1 Patient 2 Patient 3
Patient 2  4.123106                    
Patient 3 14.071247 15.842980          
Patient 4 14.594520 16.733201  1.732051</code></pre>
</div>
</div>
<p>Now the third method. correlation</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(<span class="fu">t</span>(pd))) <span class="co"># correlation distance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Patient 1   Patient 2   Patient 3
Patient 2 0.004129405                        
Patient 3 1.988522468 1.970725343            
Patient 4 1.988522468 1.970725343 0.000000000</code></pre>
</div>
</div>
<section id="an-aside-on-file-importing." class="level4">
<h4 class="anchored" data-anchor-id="an-aside-on-file-importing.">An Aside on file importing.</h4>
<p>Often I hear that one of the hardest parts about working in R is that it’s difficult to import data. I think overall it’s not as bad as they say, but I did just have a heck of a time getting a simple table imported into the format I wanted.</p>
<p>I simply made a table representing an example table from the text. I made the table in excel and saved both a .xlsx file and a .csv file. My first several attempts at reading in either of these files resulted in a variety of erros. The problem was basically that R wasn’t recognizing the first column as row names. Turns out there are a lot of functions for reading tables. And they come in different packages.</p>
<p>Here’s a few</p>
<p>read.table()</p>
<p>data.table()</p>
<p>read.csv</p>
<p>not to be confused with</p>
<p>read_csv</p>
<p>there’s also:</p>
<p>read_excel</p>
<p>read.delimit</p>
<p>the list goes on.</p>
<p>What I found worked was to save the excel file as a csv file, then use the code shown in the code chunk above. It looks like this</p>
<p>pd &lt;- read.csv(“simple_patient_data.csv”, header =TRUE, row.names= 1)</p>
<p>So, it uses read.csv an older (I think) method. The arguments are the file name, header = TRUE so it knows the first row is the header for the the data. And row.names = 1, indicating that the first column (the patient IDs in this case) should be used as row names.</p>
<p>Once I did that, the table was read properly by the various distance functions.</p>
<p>Note - I also stumbled acrosse the idea that reader functions (part of the tidyverse) don’t recognize row names. This must be part of the “tidy” organization of the data.</p>
</section>
<section id="normalizing-the-data" class="level4">
<h4 class="anchored" data-anchor-id="normalizing-the-data">Normalizing the data</h4>
<p>It is often necessary to normalize or standardize the data (see text for explanation). According to the text, a common way to normalize data is to subtract the mean and divide by the standard deviation for each value in the table. This is called “standardization”. If this is done on all genes, each gene will have the same effect on distance.</p>
<p>The function to do this in R is called scale.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scale</span>(pd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                IRX4       OCT4       PAX6
Patient 1  0.6932522  0.5212860 -1.0733721
Patient 2  1.0194886  1.1468293 -0.6214260
Patient 3 -0.7748113 -0.7298004  0.9603856
Patient 4 -0.9379295 -0.9383149  0.7344125
attr(,"scaled:center")
IRX4 OCT4 PAX6 
6.75 7.50 5.75 
attr(,"scaled:scale")
    IRX4     OCT4     PAX6 
6.130525 4.795832 4.425306 </code></pre>
</div>
</div>
<p>I suppose you would then go on to calcuate ditance on these scaled values.</p>
<p>First let’s make a new data frame pd_scaled with the scaled data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pd_scaled <span class="ot">&lt;-</span> <span class="fu">scale</span>(pd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#first the manhattan method of the dist function.  </span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dist</span>(pd_scaled,<span class="at">method=</span><span class="st">"manhattan"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Patient 1 Patient 2 Patient 3
Patient 2 1.4037257                    
Patient 3 4.7529077 5.2527412          
Patient 4 4.8985672 5.3984007 0.5976057</code></pre>
</div>
</div>
<p>Now try the Euclidean method</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dist</span>(pd_scaled,<span class="at">method=</span><span class="st">"euclidean"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Patient 1 Patient 2 Patient 3
Patient 2 0.8378483                    
Patient 3 2.8029624 3.0402926          
Patient 4 2.8388860 3.1650607 0.3480656</code></pre>
</div>
</div>
<p>Now the third method. correlation</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(<span class="fu">t</span>(pd_scaled))) <span class="co"># correlation distance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             Patient 1    Patient 2    Patient 3
Patient 2 0.0116799530                          
Patient 3 1.9978473534 1.9961863561             
Patient 4 1.9960791842 1.9979266062 0.0002634645</code></pre>
</div>
</div>
<p>Maybe later I can add plotting and a comparison of these numbers. For now we’ll leave that as an exercise for the reader :)</p>
</section>
</section>
<section id="hiearchical-clustering" class="level3">
<h3 class="anchored" data-anchor-id="hiearchical-clustering"><strong>4.1.2</strong> Hiearchical clustering</h3>
<p>HC is one of the most ubiquitous clustering algorithms. Using this algorithm you can see the relationship of individual data points and relationships of clusters. Eventually, you get a tree structure or a dendrogram that shows the relationship between the individual data points and clusters. The height of the dendrogram is the distance between clusters. The base function in R to do hierarchical clustering in&nbsp;<code>hclust()</code>. Below, we apply that function on Euclidean distances between patients. The resulting clustering tree or dendrogram is shown in Figure&nbsp;<a href="https://compgenomr.github.io/book/clustering-grouping-samples-based-on-their-similarity.html#fig:expPlot">4.1</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>d<span class="ot">=</span><span class="fu">dist</span>(pd)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>hc<span class="ot">=</span><span class="fu">hclust</span>(d,<span class="at">method=</span><span class="st">"complete"</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="compgenR_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Cool. Patients 3 and 4 cluster together and 1 and 2 form another cluster.</p>
<p>Looking at the data supports this clustering. The values are as follows</p>
<p>p1 11 10 1</p>
<p>p2 13 13 3</p>
<p>p3 2 4 10</p>
<p>p4 1 3 9</p>
<p>patients 1 and 2 are low low high</p>
<p>patients 3 are high high low.</p>
<p>So the clustering makes sense.</p>
<p>See the text for a more thorough explanation of the hclust() function and its arguments.</p>
<section id="applying-clustering-to-a-larger-data-set-and-constructing-a-heat-map." class="level4">
<h4 class="anchored" data-anchor-id="applying-clustering-to-a-larger-data-set-and-constructing-a-heat-map.">Applying clustering to a larger data set and constructing a heat map.</h4>
<p>A typical gene expression analysis comprises a much larger data set of course. One such data set contains gene expression values from 60 bone marrow samples of patients with one of the four main types of leukemia (ALL, AML, CLL, CML) or no-leukemia controls. The authors trimmed that data set down to the top 1000 most variable genes to be able to work with it more easily, since genes that are not very variable do not contribute much to the distances between patients. We will now use this data set to cluster the patients and display the values as a heatmap and a dendrogram. The heatmap function,&nbsp;pheatmap(), performs the clustering as well. The matrix that contains gene expressions has the genes in the rows and the patients in the columns. Therefore, we will also use a column-side color code to mark the patients based on their leukemia type. For the hierarchical clustering, we will use Ward’s method designated by the&nbsp;clustering_method&nbsp;argument to the&nbsp;pheatmap()&nbsp;function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load the heatmap library</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>expFile<span class="ot">=</span><span class="fu">system.file</span>(<span class="st">"extdata"</span>,<span class="st">"leukemiaExpressionSubset.rds"</span>,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">package=</span><span class="st">"compGenomRData"</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>mat<span class="ot">=</span><span class="fu">readRDS</span>(expFile)                                      </span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="co"># set the leukemia type annotation for each sample</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>annotation_col <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">LeukemiaType =</span><span class="fu">substr</span>(<span class="fu">colnames</span>(mat),<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="co">#add rownames corresponding to full sample names.  Results in a data frame with full genes as row names and one column Leukemia type with the abbreviated types.</span></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annotation_col)<span class="ot">=</span><span class="fu">colnames</span>(mat)</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="co">#next make a heatmap</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(mat,<span class="at">show_rownames=</span><span class="cn">FALSE</span>,<span class="at">show_colnames=</span><span class="cn">FALSE</span>,  <span class="co">#provide the data</span></span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col=</span>annotation_col,               <span class="co">#this is for the legend</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>         <span class="at">scale =</span> <span class="st">"none"</span>,<span class="at">clustering_method=</span><span class="st">"ward.D2"</span>,</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_distance_cols=</span><span class="st">"euclidean"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="compgenR_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>WHat is the following code doing?</p>
<p>data.frame(LeukemiaType =substr(colnames(mat),1,3)) rownames(annotation_col)=colnames(mat)</p>
<p>I believe we’re making a data frame called LekemiaType, using the first three letters of the column names of the matrix we called mat. Each column name starts with a three letter designation of the leukemia type using the substring function. In english, this does the following</p>
<p>substr(colnames(mat), 1,3) extracts the column names from mat, then takes a substring of letters1:3 for each. This is saved in a variable called annotation_col</p>
<p>the next line</p>
<p>rownames(annotation_col)=colnames(mat)</p>
<p>Changes the rownames of the annotation_col to the full column names of mat. So we get a dat frame with fullnames as rownames and abbreviations as column 1. With the column name Leukemia type.</p>
<p>Then in the pheatmap function, there is a parameter in this function called annotation_col.&nbsp;Which means, tell me what you want to use for annotation (the legend essentially). In this case we made a dataframe called annotation_col, and so we use this as the legend.</p>
</section>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>